third_party/amd/lib/TritonAMDGPUToLLVM/SchedInstructions.cpp

```diff
@@ -0,0 +1,201 @@
+  auto size = builder.create<LLVM::ConstantOp>(
+      loc, builder.getIntegerAttr(i32, sizeValue));
+  auto groupId = builder.create<LLVM::ConstantOp>(
+      loc, builder.getIntegerAttr(i32, groupIdValue));
+  builder.create<LLVM::CallIntrinsicOp>(loc, TypeRange{}, intrinsicName,
+                                        ValueRange{mask, size, groupId},
+                                        defaultFlags);
+}
+
+// Insert intrinsic that controls the types of instructions that may be
+// allowed to cross the intrinsic during instruction scheduling
+Operation *generatedSchedBarrier(PatternRewriter &rewriter, int64_t maskValue) {
+  MLIRContext *ctx = rewriter.getContext();
+  Location loc = rewriter.getUnknownLoc();
+  auto intrinsicName = StringAttr::get(ctx, "llvm.amdgcn.sched.barrier");
+  LLVM::FastmathFlagsAttr defaultFlags{};
+  Type i32 = rewriter.getI32Type();
+  auto mask = rewriter.create<LLVM::ConstantOp>(
+      loc, rewriter.getIntegerAttr(i32, maskValue));
+  return rewriter.create<LLVM::CallIntrinsicOp>(loc, TypeRange{}, intrinsicName,
+                                                ValueRange{mask}, defaultFlags);
+}
+
+// Insert an experimental intrinsic for instruction group level parallelism.
+// The intrinsic takes a value that specifies the strategy.
+Operation *generatedIglpOpt(PatternRewriter &rewriter, int value) {
+  MLIRContext *ctx = rewriter.getContext();
+  Location loc = rewriter.getUnknownLoc();
+  auto intrinsicName = StringAttr::get(ctx, "llvm.amdgcn.iglp.opt");
+  LLVM::FastmathFlagsAttr defaultFlags{};
+  Type i32 = rewriter.getI32Type();
+  auto mask = rewriter.create<LLVM::ConstantOp>(
+      loc, rewriter.getIntegerAttr(i32, static_cast<int64_t>(value)));
+  return rewriter.create<LLVM::CallIntrinsicOp>(loc, TypeRange{}, intrinsicName,
+                                                ValueRange{mask}, defaultFlags);
+}
+
+struct InstructionSchedOpRewriter
+    : public OpRewritePattern<triton::gpu::InstructionBlockSchedOp> {
+
+  InstructionSchedOpRewriter(mlir::MLIRContext *ctx, uint32_t variant)
+      : OpRewritePattern<
+            triton::gpu::InstructionBlockSchedOp>::OpRewritePattern(ctx) {
+    schedulingType = static_cast<SchedulingType>(variant);
+    schedulingType = schedulingType > SchedulingType::COUNT
+                         ? SchedulingType::NONE
+                         : schedulingType;
+  }
+
+  enum SchedulingType : uint32_t { NONE = 0, IGLP_0, IGLP_1, COUNT };
+
+  LogicalResult
+  matchAndRewrite(triton::gpu::InstructionBlockSchedOp instructionSchedOp,
+                  PatternRewriter &rewriter) const override {
+
+    bool useSchedulingGuards = true;
+    if ((schedulingType == SchedulingType::IGLP_0) ||
+        (schedulingType == SchedulingType::IGLP_1))
+      useSchedulingGuards = false;
+
+    Operation *op;
+    Block *block = instructionSchedOp->getBlock();
+    rewriter.setInsertionPointToStart(block);
```

Comment:
Sink the `setInsertionPointToStart` inside the if condition at L111? It's only needed there.

Reply:
done